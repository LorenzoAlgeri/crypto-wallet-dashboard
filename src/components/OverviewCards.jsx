import React from 'react';

const OverviewCards = ({ data }) => {
  const formatCurrency = (value) => {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 2,
      maximumFractionDigits: 2
    }).format(value);
  };

  const formatPercentage = (value) => {
    const percentage = data.totalCost > 0 ? (value / data.totalCost) * 100 : 0;
    return `${percentage >= 0 ? '+' : ''}${percentage.toFixed(2)}%`;
  };

  const cards = [
    {
      title: 'Total Value',
      value: formatCurrency(data.totalValue),
      icon: 'ðŸ’°',
      color: 'blue'
    },
    {
      title: 'Total Cost',
      value: formatCurrency(data.totalCost),
      icon: 'ðŸ“Š',
      color: 'gray'
    },
    {
      title: 'Unrealized P/L',
      value: formatCurrency(data.unrealizedPL),
      percentage: formatPercentage(data.unrealizedPL),
      icon: 'ðŸ“ˆ',
      color: data.unrealizedPL >= 0 ? 'green' : 'red',
      isProfit: data.unrealizedPL >= 0
    },
    {
      title: 'Realized P/L',
      value: formatCurrency(data.realizedPL),
      icon: 'ðŸ’¸',
      color: data.realizedPL >= 0 ? 'green' : 'red',
      isProfit: data.realizedPL >= 0
    }
  ];

  const getColorClasses = (color, isProfit) => {
    const colorMap = {
      blue: 'bg-blue-50 border-blue-200',
      gray: 'bg-gray-50 border-gray-200',
      green: 'bg-green-50 border-green-200',
      red: 'bg-red-50 border-red-200'
    };

    const iconColorMap = {
      blue: 'text-blue-600',
      gray: 'text-gray-600',
      green: 'text-green-600',
      red: 'text-red-600'
    };

    return {
      background: colorMap[color],
      icon: iconColorMap[color]
    };
  };

  return (
    <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-6">
      {cards.map((card, index) => {
        const colors = getColorClasses(card.color, card.isProfit);
        
        return (
          <div
            key={index}
            className={`bg-white rounded-lg shadow-sm border p-6 ${colors.background}`}
          >
            <div className="flex items-center justify-between">
              <div className="flex items-center space-x-3">
                <div className={`w-10 h-10 rounded-lg flex items-center justify-center ${colors.background}`}>
                  <span className={`text-xl ${colors.icon}`}>{card.icon}</span>
                </div>
                <div>
                  <p className="text-sm font-medium text-gray-600">{card.title}</p>
                  <p className={`text-2xl font-bold ${
                    card.color === 'green' ? 'text-green-600' : 
                    card.color === 'red' ? 'text-red-600' : 'text-gray-900'
                  }`}>
                    {card.value}
                  </p>
                  {card.percentage && (
                    <p className={`text-sm font-medium ${
                      card.isProfit ? 'text-green-600' : 'text-red-600'
                    }`}>
                      {card.percentage}
                    </p>
                  )}
                </div>
              </div>
              {card.isProfit !== undefined && (
                <div className={`text-2xl ${card.isProfit ? 'text-green-500' : 'text-red-500'}`}>
                  {card.isProfit ? 'ðŸ“ˆ' : 'ðŸ“‰'}
                </div>
              )}
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default OverviewCards;
