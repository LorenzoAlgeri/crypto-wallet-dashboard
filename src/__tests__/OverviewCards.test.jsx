import React from 'react';
import { render } from '@testing-library/react';
import OverviewCards from '../components/OverviewCards.jsx';

describe('OverviewCards', () => {
  const mockData = {
    totalValue: 10000,
    totalCost: 8000,
    unrealizedPL: 2000,
    realizedPL: 500
  };

  const mockDataNegative = {
    totalValue: 6000,
    totalCost: 8000,
    unrealizedPL: -2000,
    realizedPL: -500
  };

  const mockDataZero = {
    totalValue: 0,
    totalCost: 0,
    unrealizedPL: 0,
    realizedPL: 0
  };

  it('renders correctly with positive P/L data', () => {
    const { container } = render(<OverviewCards data={mockData} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with negative P/L data', () => {
    const { container } = render(<OverviewCards data={mockDataNegative} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with zero values', () => {
    const { container } = render(<OverviewCards data={mockDataZero} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays correct formatted currency values', () => {
    const { getByText } = render(<OverviewCards data={mockData} />);
    
    expect(getByText('$10,000.00')).toBeInTheDocument();
    expect(getByText('$8,000.00')).toBeInTheDocument();
    expect(getByText('$2,000.00')).toBeInTheDocument();
    expect(getByText('$500.00')).toBeInTheDocument();
  });

  it('displays correct percentage for unrealized P/L', () => {
    const { getByText } = render(<OverviewCards data={mockData} />);
    
    // 2000 / 8000 * 100 = 25%
    expect(getByText('+25.00%')).toBeInTheDocument();
  });

  it('displays correct icons and colors for positive values', () => {
    const { container } = render(<OverviewCards data={mockData} />);
    
    // Check for profit indicators
    const profitEmojis = container.querySelectorAll('[class*="text-green"]');
    expect(profitEmojis.length).toBeGreaterThan(0);
    
    // Check for trend up emoji
    expect(container.textContent).toContain('ðŸ“ˆ');
  });

  it('displays correct icons and colors for negative values', () => {
    const { container } = render(<OverviewCards data={mockDataNegative} />);
    
    // Check for loss indicators
    const lossEmojis = container.querySelectorAll('[class*="text-red"]');
    expect(lossEmojis.length).toBeGreaterThan(0);
    
    // Check for trend down emoji
    expect(container.textContent).toContain('ðŸ“‰');
  });

  it('handles zero total cost correctly for percentage calculation', () => {
    const { container } = render(<OverviewCards data={mockDataZero} />);
    
    // Should not throw error and should display 0.00%
    expect(container.textContent).toContain('+0.00%');
  });

  it('renders all four cards', () => {
    const { container } = render(<OverviewCards data={mockData} />);
    
    const cards = container.querySelectorAll('[class*="rounded-lg shadow-sm border p-6"]');
    expect(cards).toHaveLength(4);
  });

  it('displays correct card titles', () => {
    const { getByText } = render(<OverviewCards data={mockData} />);
    
    expect(getByText('Total Value')).toBeInTheDocument();
    expect(getByText('Total Cost')).toBeInTheDocument();
    expect(getByText('Unrealized P/L')).toBeInTheDocument();
    expect(getByText('Realized P/L')).toBeInTheDocument();
  });
});
