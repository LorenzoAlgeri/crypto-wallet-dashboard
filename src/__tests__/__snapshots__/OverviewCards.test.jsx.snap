// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewCards renders correctly with negative P/L data 1`] = `
<div
  class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-6"
>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-blue-50 border-blue-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-blue-50 border-blue-200"
        >
          <span
            class="text-xl text-blue-600"
          >
            ðŸ’°
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Total Value
          </p>
          <p
            class="text-2xl font-bold text-gray-900"
          >
            $6,000.00
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-gray-50 border-gray-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-gray-50 border-gray-200"
        >
          <span
            class="text-xl text-gray-600"
          >
            ðŸ“Š
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Total Cost
          </p>
          <p
            class="text-2xl font-bold text-gray-900"
          >
            $8,000.00
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-red-50 border-red-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-red-50 border-red-200"
        >
          <span
            class="text-xl text-red-600"
          >
            ðŸ“ˆ
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Unrealized P/L
          </p>
          <p
            class="text-2xl font-bold text-red-600"
          >
            -$2,000.00
          </p>
          <p
            class="text-sm font-medium text-red-600"
          >
            -25.00%
          </p>
        </div>
      </div>
      <div
        class="text-2xl text-red-500"
      >
        ðŸ“‰
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-red-50 border-red-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-red-50 border-red-200"
        >
          <span
            class="text-xl text-red-600"
          >
            ðŸ’¸
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Realized P/L
          </p>
          <p
            class="text-2xl font-bold text-red-600"
          >
            -$500.00
          </p>
        </div>
      </div>
      <div
        class="text-2xl text-red-500"
      >
        ðŸ“‰
      </div>
    </div>
  </div>
</div>
`;

exports[`OverviewCards renders correctly with positive P/L data 1`] = `
<div
  class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-6"
>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-blue-50 border-blue-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-blue-50 border-blue-200"
        >
          <span
            class="text-xl text-blue-600"
          >
            ðŸ’°
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Total Value
          </p>
          <p
            class="text-2xl font-bold text-gray-900"
          >
            $10,000.00
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-gray-50 border-gray-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-gray-50 border-gray-200"
        >
          <span
            class="text-xl text-gray-600"
          >
            ðŸ“Š
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Total Cost
          </p>
          <p
            class="text-2xl font-bold text-gray-900"
          >
            $8,000.00
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-green-50 border-green-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-green-50 border-green-200"
        >
          <span
            class="text-xl text-green-600"
          >
            ðŸ“ˆ
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Unrealized P/L
          </p>
          <p
            class="text-2xl font-bold text-green-600"
          >
            $2,000.00
          </p>
          <p
            class="text-sm font-medium text-green-600"
          >
            +25.00%
          </p>
        </div>
      </div>
      <div
        class="text-2xl text-green-500"
      >
        ðŸ“ˆ
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-green-50 border-green-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-green-50 border-green-200"
        >
          <span
            class="text-xl text-green-600"
          >
            ðŸ’¸
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Realized P/L
          </p>
          <p
            class="text-2xl font-bold text-green-600"
          >
            $500.00
          </p>
        </div>
      </div>
      <div
        class="text-2xl text-green-500"
      >
        ðŸ“ˆ
      </div>
    </div>
  </div>
</div>
`;

exports[`OverviewCards renders correctly with zero values 1`] = `
<div
  class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-6"
>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-blue-50 border-blue-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-blue-50 border-blue-200"
        >
          <span
            class="text-xl text-blue-600"
          >
            ðŸ’°
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Total Value
          </p>
          <p
            class="text-2xl font-bold text-gray-900"
          >
            $0.00
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-gray-50 border-gray-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-gray-50 border-gray-200"
        >
          <span
            class="text-xl text-gray-600"
          >
            ðŸ“Š
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Total Cost
          </p>
          <p
            class="text-2xl font-bold text-gray-900"
          >
            $0.00
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-green-50 border-green-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-green-50 border-green-200"
        >
          <span
            class="text-xl text-green-600"
          >
            ðŸ“ˆ
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Unrealized P/L
          </p>
          <p
            class="text-2xl font-bold text-green-600"
          >
            $0.00
          </p>
          <p
            class="text-sm font-medium text-green-600"
          >
            +0.00%
          </p>
        </div>
      </div>
      <div
        class="text-2xl text-green-500"
      >
        ðŸ“ˆ
      </div>
    </div>
  </div>
  <div
    class="bg-white rounded-lg shadow-sm border p-6 bg-green-50 border-green-200"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg flex items-center justify-center bg-green-50 border-green-200"
        >
          <span
            class="text-xl text-green-600"
          >
            ðŸ’¸
          </span>
        </div>
        <div>
          <p
            class="text-sm font-medium text-gray-600"
          >
            Realized P/L
          </p>
          <p
            class="text-2xl font-bold text-green-600"
          >
            $0.00
          </p>
        </div>
      </div>
      <div
        class="text-2xl text-green-500"
      >
        ðŸ“ˆ
      </div>
    </div>
  </div>
</div>
`;
